include/ofi_list.h

```diff
@@ -457,6 +468,17 @@ static inline struct slist_entry *slist_remove_head(struct slist *list)
 
 typedef int slist_func_t(struct slist_entry *item, const void *arg);
 
+static inline size_t slist_size(struct slist *list)
+{
+	size_t size = 0;
+	struct slist_entry *cur, *prev;
+
+	slist_foreach(list, cur, prev) {
+		size++;
+	}
+	return size;
+}
```

Comment:
We probably want both of these to be defined only for debug builds.  The implementation is slow, and we want to make sure someone doesn't rely on these function in production code, assuming that they were implemented efficiently.

Reply:
On further thought, instead of making these calls debug only, I'd add a debug size variable that's incremented/decremented on list add/remove.  That way, the overhead of calling size() is minimized during debug builds, which will allow it to be checked more frequently.  See OFI_DGB_VAR/ADD in ofi_osd.h.

Reply:
Where should this size variable go? It could go on the head of the list or on every entry. But updating would be messy in both cases with entry removals and splicing.

Reply:
Uhm, yeah, this would be easy for slist.  Not so much for dlist, depending on how dlist is used.

Looking at dlist, we could definitely use some debug level tracking to help catch bugs.  This may be more work that you're wanting to do now.  I wonder if we could start with something like this:
```
struct dlist_entry {
        struct dlist_entry      *next;
        struct dlist_entry      *prev;
+
+       struct dlist_entry      *list;
 };
 
 #define DLIST_INIT(addr) { addr, addr }
@@ -66,6 +68,8 @@ static inline void dlist_init(struct dlist_entry *head)
 {
        head->next = head;
        head->prev = head;
+
+       head->list = head;
 }
 
 static inline int dlist_empty(struct dlist_entry *head)
@@ -76,6 +80,10 @@ static inline int dlist_empty(struct dlist_entry *head)
 static inline void
 dlist_insert_after(struct dlist_entry *item, struct dlist_entry *head)
 {
+       assert(!item->list || item->list == item);
+       assert(head->list);
+       item->list = head->list;
+
        item->next = head->next;
        item->prev = head;
        head->next->prev = item;
@@ -93,6 +101,9 @@ dlist_insert_before(struct dlist_entry *item, struct dlist_entry *he
 
 static inline void dlist_remove(struct dlist_entry *item)
 {
+       assert(item->list);
+       item->list = NULL;
+
        item->prev->next = item->next;
        item->next->prev = item->prev;
 }
```
The list accesses should be debug only, and I only spent about 5 minutes adding this.  I could easily have missed something, plus I didn't try to update all the calls.  Anyway, the goal is to have each entry reference the head of the list that the entry is in.  If we can add this tracking, then adding a size to the head of the list is doable.  A spliced list would sum the 2 sizes together.

Reply:
@shefty can you take a look at this version?